text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,633 @@
+deduplicating each of the areas above. While all of these new features are
+pretty small in their own right, they all add up to greatly reducing the
+overhead of managing a workspace of many crates. The list of new features in
+Cargo will look like the following:
+
+## Workspace-level Dependencies
+
+The `[workspace]` section can now have a `dependencies` section which works the
+same way as the `[dependencies]` section in `Cargo.toml`:
+
+```toml
+# in workspace's Cargo.toml
+[workspace.dependencies]
+log = "0.3.1"
+log2 = { version = "2.0.0", package = "log" }
+serde = { git = 'https://github.com/serde-rs/serde" }
+wasm-bindgen-cli = { path = "crates/cli" }
+```
+
+Each workspace member can then reference this section in the workspace with a
+new dependency directive:
+
+```toml
+# in a workspace member's Cargo.toml
+[dependencies]
+log = { workspace = true }
+```
+
+This directive indicates that the `log` dependency should be looked up from
+`workspace.dependencies` in the workspace root. You can also indicate which
+dependency should be used with:
+
+```toml
+[dependencies]
+log = { workspace = "log2" }
+```
+
+where this indicates that the crate's `log` dependency will be the `log2`
+dependency specified in the workspace `Cargo.toml`. This means that to this
+local crate `log` will be `log v2.0.0` from crates.io.
+
+## No longer need both `version` and `path` to publish to Crates.io
+
+When you have a `path` dependency, Cargo's current behavior on publication looks
+like this:
+
+* If you have a `version` specifier as well, then the `path` key is deleted and
+  the crate is uploaded with the specified `version` as a dependency
+  requirement.
+* If you don't have a `version` specifier, then the dependency directive is
+  deleted and crates.io will not learn about this dependency. This is only
+  really useful for `dev-dependencies`.
+
+Cargo's behavior will change in this second case, instead following new logic
+for a missing `version` specifier. For dev-dependencies where the referenced
+package is `publish = false`, then the dependency will be dropped. Otherwise
+Cargo will assume that `version = "$dependency_version"` was specified, meaning
+that it requires at least the current version and otherwise any
+semver-compatible version.
+
+This behavior should mean that you no longer need to write `version = "..."`
+with `path` dependencies if you publish to crates.io. Coupled with the
+workspace-level dependencies above this means you never have to write the
+version of a path dependency anywhere!
+
+## Package versions can reference other workspace members
+
+To deduplicate `version` directives in `Cargo.toml` workspace members, Cargo
+will now support a way to indirectly declare the version of your package. The
+way this works is that you say a package's version should be the same as
+another's, like so:
+
+```toml
+[package]
+name = "foo"
+version = { workspace = "other-workspace-member" }
```

Comment:
Why is version handled differently from authors, license, etc?

Reply:
That's not a bad point actually, I'll update so everything can uniformly reference other workspace members, be inherited, or be written inline.